transports/webtransport-websys/src/transport.rs

```diff
@@ -0,0 +1,191 @@
+use futures::future::FutureExt;
+use libp2p_core::muxing::StreamMuxerBox;
+use libp2p_core::transport::{Boxed, ListenerId, Transport as _, TransportError, TransportEvent};
+use libp2p_identity::{Keypair, PeerId};
+use multiaddr::Multiaddr;
+use send_wrapper::SendWrapper;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::ready;
+use std::task::{Context, Poll};
+
+use crate::connection::ConnectionSend;
+use crate::endpoint::Endpoint;
+use crate::Connection;
+use crate::Error;
+
+/// Config for the [`Transport`].
+pub struct Config {
+    keypair: Keypair,
+}
+
+/// A WebTransport [`Transport`](libp2p_core::Transport) that works with `web-sys`.
+pub struct Transport {
+    config: Config,
+}
+
+/// Transport wrapped in [`SendWrapper`].
+///
+/// This is needed by Swarm. WASM is single-threaded and it is safe
+/// to use [`SendWrapper`].
+pub(crate) struct TransportSend {
+    inner: SendWrapper<Transport>,
+}
+
+impl Config {
+    /// Constructs a new configuration for the [`Transport`].
+    pub fn new(keypair: &Keypair) -> Self {
+        Config {
+            keypair: keypair.to_owned(),
+        }
+    }
+}
+
+impl Transport {
+    /// Constructs a new `Transport` with the given [`Config`].
+    pub fn new(config: Config) -> Transport {
+        Transport { config }
+    }
+
+    /// Wraps `Transport` in [`Boxed`] and makes it ready to be consumed by
+    /// SwarmBuilder.
+    pub fn boxed(self) -> Boxed<(PeerId, StreamMuxerBox)> {
+        TransportSend::new(self)
+            .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+            .boxed()
+    }
+}
+
+impl libp2p_core::Transport for Transport {
+    type Output = (PeerId, Connection);
+    type Error = Error;
+    type ListenerUpgrade = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>>>>;
+    type Dial = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>>>>;
+
+    fn listen_on(
+        &mut self,
+        _id: ListenerId,
+        addr: Multiaddr,
+    ) -> Result<(), TransportError<Self::Error>> {
+        Err(TransportError::MultiaddrNotSupported(addr))
+    }
+
+    fn remove_listener(&mut self, _id: ListenerId) -> bool {
+        false
+    }
+
+    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let endpoint = Endpoint::from_multiaddr(&addr).map_err(|e| match e {
+            e @ Error::InvalidMultiaddr(_) => {
+                log::error!("{}", e);
```

Comment:
I don't think this should be an error, it can happen regularly when an address is passed down the chain of transports.

I do agree that debugging this is hard. We should figure out how to add useful logs for this here: https://github.com/libp2p/rust-libp2p/blob/14530af2611b0cb17179f83802d31c0f28222e91/core/src/transport/choice.rs#L49-L89

Reply:
I opened an issue: https://github.com/libp2p/rust-libp2p/issues/4072

Reply:
I am happy to keep the log but `error!` is too extreme. The rule of thumb for `warn` and `error` that I always apply is: "Do I want to wake an ops-person at 3am with this log?"

Reply:
I though you didn't want the log in general :). Sorry for the misunderstanding, I can change it to `warn`

Reply:
Done